File: lib/src/main/java/com/orientsec/easysocket/task/LifecycleCallbackWrapper.java
package com.orientsec.easysocket.task;

import androidx.annotation.NonNull;

import com.orientsec.easysocket.Options;
import com.orientsec.easysocket.Packet;
import com.orientsec.easysocket.client.BaseSocketClient;
import com.orientsec.easysocket.utils.Logger;

import java.util.concurrent.Executor;

public class LifecycleCallbackWrapper<T> implements LifecycleCallback<T> {

    private final Callback<T> callback;
    private final Executor executor;
    private final Logger logger;
    private final Task<?> task;
    private final boolean isDebuggable;

    public LifecycleCallbackWrapper(Callback<T> callback, Task<?> task, BaseSocketClient client) {
        this.callback = callback;
        this.logger = client.getLogger();
        this.task = task;
        Options options = client.getOptions();
        this.executor = options.getCallbackExecutor();
        this.isDebuggable = options.isDebuggable();
    }

    @Override
    public void onStart() {
        if (isDebuggable) logger.d(prefix() + "onStart");
        executor.execute(callback::onStart);
    }

    @Override
    public void onSuccess(@NonNull T res) {
        if (isDebuggable)
            logger.d(prefix() + "onSuccess, type: " + res.getClass().getSimpleName());
        executor.execute(() -> {
            callback.onSuccess(res);
            if (callback instanceof LifecycleCallback) {
                ((LifecycleCallback<T>) callback).onComplete();
            }
        });
        onComplete();
    }

    @Override
    public void onFailure(@NonNull Throwable t) {
        if (isDebuggable) logger.d(prefix() + "onFailure, error: " + t.getMessage());
        executor.execute(() -> {
            callback.onFailure(t);
            if (callback instanceof LifecycleCallback) {
                ((LifecycleCallback<T>) callback).onComplete();
            }
        });
        onComplete();
    }

    @Override
    public void onCanceled() {
        if (isDebuggable) logger.d(prefix() + "onCanceled");
        executor.execute(() -> {
            callback.onCanceled();
            if (callback instanceof LifecycleCallback) {
                ((LifecycleCallback<T>) callback).onComplete();
            }
        });
        onComplete();
    }

    @Override
    public void onWait() {
        if (isDebuggable) logger.d(prefix() + "onWait");
        if (callback instanceof LifecycleCallback) {
            executor.execute(((LifecycleCallback<T>) callback)::onWait);
        }
    }

    @Override
    public void onResume() {
        if (isDebuggable) logger.d(prefix() + "onResume");
        if (callback instanceof LifecycleCallback) {
            executor.execute(((LifecycleCallback<T>) callback)::onResume);
        }
    }

    @Override
    public void onEncodeStart() {
        if (isDebuggable) logger.d(prefix() + "onEncodeStart");
        if (callback instanceof LifecycleCallback) {
            executor.execute(((LifecycleCallback<T>) callback)::onEncodeStart);
        }
    }

    @Override
    public void onEncodeSuccess() {
        if (isDebuggable) logger.d(prefix() + "onEncodeSuccess");
        if (callback instanceof LifecycleCallback) {
            executor.execute(((LifecycleCallback<T>) callback)::onEncodeSuccess);
        }
    }

    @Override
    public void onEncodeFailure(Throwable t) {
        if (isDebuggable) logger.d(prefix() + "onEncodeFailure, error: " + t.getMessage());
        if (callback instanceof LifecycleCallback) {
            executor.execute(() -> ((LifecycleCallback<T>) callback).onEncodeFailure(t));
        }
    }

    @Override
    public void onReset(int failedTimes, Throwable t) {
        if (isDebuggable)
            logger.d(prefix() + "onReset, failedTimes: " + failedTimes
                    + ", error: " + t.getMessage());
        if (callback instanceof LifecycleCallback) {
            executor.execute(() -> ((LifecycleCallback<T>) callback).onReset(failedTimes, t));
        }
    }

    @Override
    public void onSendStart() {
        if (isDebuggable) logger.d(prefix() + "onSendStart");
        if (callback instanceof LifecycleCallback) {
            executor.execute(((LifecycleCallback<T>) callback)::onSendStart);
        }
    }

    @Override
    public void onSendSuccess() {
        if (isDebuggable) logger.d(prefix() + "onSendSuccess");
        if (callback instanceof LifecycleCallback) {
            executor.execute(((LifecycleCallback<T>) callback)::onSendSuccess);
        }
    }

    @Override
    public void onSendFailure(Throwable t) {
        if (isDebuggable) logger.d(prefix() + "onSendFailure, error: " + t.getMessage());
        if (callback instanceof LifecycleCallback) {
            executor.execute(() -> ((LifecycleCallback<T>) callback).onSendFailure(t));
        }
    }

    @Override
    public void onPacketReceived(Packet packet) {
        if (isDebuggable) logger.d(prefix() + "onPacketReceived, packet: " + packet);
        if (callback instanceof LifecycleCallback) {
            executor.execute(() -> ((LifecycleCallback<T>) callback).onPacketReceived(packet));
        }
    }

    @Override
    public void onDecodeStart() {
        if (isDebuggable) logger.d(prefix() + "onDecodeStart");
        if (callback instanceof LifecycleCallback) {
            executor.execute(((LifecycleCallback<T>) callback)::onDecodeStart);
        }
    }

    @Override
    public void onDecodeSuccess() {
        if (isDebuggable) logger.d(prefix() + "onDecodeSuccess");
        if (callback instanceof LifecycleCallback) {
            executor.execute(((LifecycleCallback<T>) callback)::onDecodeSuccess);
        }
    }

    @Override
    public void onDecodeFailure(Throwable t) {
        if (isDebuggable) logger.d(prefix() + "onDecodeFailure, error: " + t.getMessage());
        if (callback instanceof LifecycleCallback) {
            executor.execute(() -> ((LifecycleCallback<T>) callback).onDecodeFailure(t));
        }
    }

    @Override
    public void onComplete() {
        if (isDebuggable) logger.d(prefix() + "onComplete");
    }

    private String prefix() {
        return "Task " + task.getTaskId() + ", type: " + task.getTaskType() + ", ";
    }
}

File: lib/src/main/kotlin/com/orientsec/easysocket/task/LifecycleCallbackWrapper.kt
package com.orientsec.easysocket.task

import com.orientsec.easysocket.Packet
import com.orientsec.easysocket.client.BaseSocketClient
import com.orientsec.easysocket.utils.Logger
import java.util.concurrent.Executor

class LifecycleCallbackWrapper<T>(
    private val callback: Callback<T>,
    private val task: Task<*>,
    client: BaseSocketClient
) : LifecycleCallback<T> {

    private val executor: Executor
    private val logger: Logger = client.logger
    private val isDebuggable: Boolean

    init {
        val options = client.options
        this.executor = options.callbackExecutor
        this.isDebuggable = options.isDebuggable
    }

    override fun onStart() {
        if (isDebuggable) logger.d(prefix() + "onStart")
        executor.execute { callback.onStart() }
    }

    override fun onSuccess(res: T) {
        if (isDebuggable)
            logger.d(prefix() + "onSuccess, type: " + res!!::class.java.simpleName)
        executor.execute {
            callback.onSuccess(res)
            (callback as? LifecycleCallback<T>)?.onComplete()
        }
        onComplete()
    }

    override fun onFailure(t: Throwable) {
        if (isDebuggable) logger.d(prefix() + "onFailure, error: " + t.message)
        executor.execute {
            callback.onFailure(t)
            (callback as? LifecycleCallback<T>)?.onComplete()
        }
        onComplete()
    }

    override fun onCanceled() {
        if (isDebuggable) logger.d(prefix() + "onCanceled")
        executor.execute {
            callback.onCanceled()
            (callback as? LifecycleCallback<T>)?.onComplete()
        }
        onComplete()
    }

    override fun onWait() {
        if (isDebuggable) logger.d(prefix() + "onWait")
        (callback as? LifecycleCallback<T>)?.let { lifecycleCallback ->
            executor.execute { lifecycleCallback.onWait() }
        }
    }

    override fun onResume() {
        if (isDebuggable) logger.d(prefix() + "onResume")
        (callback as? LifecycleCallback<T>)?.let { lifecycleCallback ->
            executor.execute { lifecycleCallback.onResume() }
        }
    }

    override fun onEncodeStart() {
        if (isDebuggable) logger.d(prefix() + "onEncodeStart")
        (callback as? LifecycleCallback<T>)?.let { lifecycleCallback ->
            executor.execute { lifecycleCallback.onEncodeStart() }
        }
    }

    override fun onEncodeSuccess() {
        if (isDebuggable) logger.d(prefix() + "onEncodeSuccess")
        (callback as? LifecycleCallback<T>)?.let { lifecycleCallback ->
            executor.execute { lifecycleCallback.onEncodeSuccess() }
        }
    }

    override fun onEncodeFailure(t: Throwable) {
        if (isDebuggable) logger.d(prefix() + "onEncodeFailure, error: " + t.message)
        (callback as? LifecycleCallback<T>)?.let { lifecycleCallback ->
            executor.execute { lifecycleCallback.onEncodeFailure(t) }
        }
    }

    override fun onReset(failedTimes: Int, t: Throwable) {
        if (isDebuggable)
            logger.d(
                prefix() + "onReset, failedTimes: " + failedTimes
                        + ", error: " + t.message
            )
        (callback as? LifecycleCallback<T>)?.let { lifecycleCallback ->
            executor.execute { lifecycleCallback.onReset(failedTimes, t) }
        }
    }

    override fun onSendStart() {
        if (isDebuggable) logger.d(prefix() + "onSendStart")
        (callback as? LifecycleCallback<T>)?.let { lifecycleCallback ->
            executor.execute { lifecycleCallback.onSendStart() }
        }
    }

    override fun onSendSuccess() {
        if (isDebuggable) logger.d(prefix() + "onSendSuccess")
        (callback as? LifecycleCallback<T>)?.let { lifecycleCallback ->
            executor.execute { lifecycleCallback.onSendSuccess() }
        }
    }

    override fun onSendFailure(t: Throwable) {
        if (isDebuggable) logger.d(prefix() + "onSendFailure, error: " + t.message)
        (callback as? LifecycleCallback<T>)?.let { lifecycleCallback ->
            executor.execute { lifecycleCallback.onSendFailure(t) }
        }
    }

    override fun onPacketReceived(packet: Packet) {
        if (isDebuggable) logger.d(prefix() + "onPacketReceived, packet: " + packet)
        (callback as? LifecycleCallback<T>)?.let { lifecycleCallback ->
            executor.execute { lifecycleCallback.onPacketReceived(packet) }
        }
    }

    override fun onDecodeStart() {
        if (isDebuggable) logger.d(prefix() + "onDecodeStart")
        (callback as? LifecycleCallback<T>)?.let { lifecycleCallback ->
            executor.execute { lifecycleCallback.onDecodeStart() }
        }
    }

    override fun onDecodeSuccess() {
        if (isDebuggable) logger.d(prefix() + "onDecodeSuccess")
        (callback as? LifecycleCallback<T>)?.let { lifecycleCallback ->
            executor.execute { lifecycleCallback.onDecodeSuccess() }
        }
    }

    override fun onDecodeFailure(t: Throwable) {
        if (isDebuggable) logger.d(prefix() + "onDecodeFailure, error: " + t.message)
        (callback as? LifecycleCallback<T>)?.let { lifecycleCallback ->
            executor.execute { lifecycleCallback.onDecodeFailure(t) }
        }
    }

    override fun onComplete() {
        if (isDebuggable) logger.d(prefix() + "onComplete")
    }

    private fun prefix(): String {
        return "Task " + task.taskId + ", type: " + task.taskType + ", "
    }
}
